src/vulkan/engine_vulkan.cc

```diff
@@ -244,8 +244,61 @@ Result EngineVulkan::DoClear(const ClearCommand*) {
   return pipeline_->AsGraphics()->Clear();
 }
 
-Result EngineVulkan::DoDrawRect(const DrawRectCommand*) {
-  return Result("Vulkan::DoDrawRect Not Implemented");
+Result EngineVulkan::DoDrawRect(const DrawRectCommand* command) {
+  if (!pipeline_->IsGraphics())
+    return Result("Vulkan::DrawRect for Non-Graphics Pipeline");
+
+  auto* graphics = pipeline_->AsGraphics();
+
+  Result r = graphics->ResetPipelineAndVertexBuffer();
+  if (!r.IsSuccess())
+    return r;
+
+  Format format;
+  format.SetFormatType(FormatType::kR32G32_SFLOAT);
+  format.AddComponent(FormatComponentType::kR, FormatMode::kSFloat, 32);
+  format.AddComponent(FormatComponentType::kG, FormatMode::kSFloat, 32);
+
+  float x = command->GetX();
+  float y = command->GetY();
+  float width = command->GetWidth();
+  float height = command->GetHeight();
+  if (command->IsOrtho()) {
+    const float frame_width = static_cast<float>(graphics->GetWidth());
+    const float frame_height = static_cast<float>(graphics->GetHeight());
+    x = ((x / frame_width) * 2.0f) - 1.0f;
+    y = ((y / frame_height) * 2.0f) - 1.0f;
+    width = ((width / frame_width) * 2.0f) - 1.0f;
+    height = ((height / frame_height) * 2.0f) - 1.0f;
+  }
+
+  std::vector<Value> values(8);
+  // Bottom left
+  values[0].SetDoubleValue(static_cast<double>(x));
```

Comment:
The values are floats, and they should automatically promote to doubles. So I think you can (and should) skip the static casts.

Reply:
@dj2 says GCC's -Wdouble-promotion will issue a warning if the static casts are removed.  Ok then.  I find that sad.
An alternate is to make the x, y, width, height variables doubles.. But that's probably overkill.